kv/dist_sender.go

```diff
@@ -396,6 +403,15 @@ func (ds *DistSender) Send(call *client.Call) {
 				}
 			} else if isMulti {
 				// If this request spans ranges, collect the replies.
+				if maxResults > 0 {
+					if reply, ok := reply.(proto.Truncatable); ok {
```

Comment:
Maybe we can manage to do all the multi-range work here, so you could Combine() here, then do a RowCount() (if applicable) and stop quering more ranges (or truncate & break if you're above maxRetries)
Ideally that should be done without allocating extra variables when the request does not span ranges.

I also think there are some bugs hidden here due to `gogoproto.Merge` (see https://github.com/cockroachdb/cockroach/issues/494) without properly resetting. Feel free to fix those when you spot them.